prelude/slang-cuda-prelude.h

```diff
@@ -1418,10 +1419,16 @@ __inline__ __device__ bool _waveIsSingleLane(WarpMask mask)
 }
 
 // Returns the power of 2 size of run of set bits. Returns 0 if not a suitable run.
+// Examples:
+// 0b00000000'00000000'00000000'11111111 -> 8
+// 0b11111111'11111111'11111111'11111111 -> 32
+// 0b00000000'00000000'00000000'00011111 -> 0 (since 5 is not a power of 2)
+// 0b00000000'00000000'00000000'11110000 -> 0 (since the run of bits does not start at the LSB)
+// 0b00000000'00000000'00000000'00100111 -> 0 (since it is not a single contiguous run)
 __inline__ __device__ int _waveCalcPow2Offset(WarpMask mask)
 {
     // This should be the most common case, so fast path it
-    if (mask == SLANG_CUDA_WARP_MASK)
+    if (mask == SLANG_CUDA_WARP_BITMASK)
```

Comment:
Could you explain a little more about this change? Why would the mask be ~int(0)? 

Would it be better to do `if (mask &  SLANG_CUDA_WARP_MASK ) == SLANG_CUDA_WARP_MASK`, which would cover scenarios where user code set the mask to more bits than lanes?

Reply:
Absolutely! This code isn't totally portable at the moment, because it relies on `typedef int WarpMask` being exactly `#define SLANG_CUDA_WARP_SIZE 32` bits long. (Because each bit in `WarpMask` corresponds to an active thread, we want every bit in WarpMask to be set; this is why `~int(0)` should work for the 32-bit, 32-thread case.)

If we can assume `SLANG_CUDA_WARP_SIZE` is always 32, then maybe I could change the type of `WarpMask` to `int32_t` (which could protect against the case where `int` is a length other than 32 bits on some platforms)?

In the most general case, the tricky thing is that trying to set only the least significant `SLANG_CUDA_WARP_SIZE` bits using

```cpp
#define SLANG_CUDA_WARP_BITMASK (1 << SLANG_CUDA_WARP_SIZE) - 1
```

invokes undefined behavior, because it shifts left by the number of bits in the type: from the very end of "built-in bitwise shift operators" in https://en.cppreference.com/w/cpp/language/operator_arithmetic,

> If the value of rhs is negative or is not less than the number of bits in lhs, the behavior is undefined.

If this winds up in host code, it turns out `clang -O2` recognizes this UB and totally optimizes it out, which leads to surprising behavior (this Godbolt sample shows it for unsigned ints, but the same UB happens for signed types): https://godbolt.org/z/s4TG5Wzdj

This would probably work, but it's pretty long and it could be nice if there's a nice way to do it at the preprocessor level -- what do you think about this?

```cpp
#define SLANG_CUDA_WARP_BITMASK ((SLANG_CUDA_WARP_SIZE >= sizeof(WarpMask)) ? ~WarpMask(0) : (WarpMask(1) << SLANG_CUDA_WARP_SIZE) - 1)
```

Reply:
Oh, I should probably describe more how `_waveCalcPow2Offset(mask)` gets called with a mask of active threads, rather than the number of active threads. For a scalar type, `WaveActiveSum<T>(x)` compiles to `_waveSum(_getActiveMask(), x)`, and then the return value of  `_getActiveMask()` gets passed to `_waveCalcPow2Offset(mask)` via the following call stack:

```cpp
_waveSum<T>(_getActiveMask(), x):
  _waveReduceScalar<WaveOpAdd<T>, T>(mask, val):
    const int offsetSize = _waveCalcPow2Offset(mask)
```

Thanks!

Reply:
Ok I see, that makes a lot of sense. I think your change as it seems reasonable and agree

```
#define SLANG_CUDA_WARP_BITMASK ((SLANG_CUDA_WARP_SIZE >= sizeof(WarpMask)) ? ~WarpMask(0) : (WarpMask(1) << SLANG_CUDA_WARP_SIZE) - 1)
```

is perhaps overkill.